### ISSUE TITLE
[Feat] Add Grounding Metrics to Vipassana for OOD/Hallucination Detection

### ISSUE THREAD

**Issue by tamatebox** [tamatebox/samadhi_model#1]
## Summary

Vipassanaの現在の実装では、Trust Scoreの計算に使用する`min_dist`（Probeへの最小距離）をS*（収束後の状態）から計算している。しかし、Vicaraの収束特性により、ID（In-Distribution）とOOD（Out-of-Distribution）の入力が同じS*領域に引き寄せられてしまい、OOD検出が機能しない。

### 現象
- Trust Scoreのヒートマップが「円形」パターンではなく「水平縞」パターンになる
- ID/OOD間でmin_distがほぼ同一（比率: 1.006）
- ほぼ全てのサンプルが同じProbe（Probe 4または6）に収束

### 根本原因
Vicaraの設計上、あらゆる入力を既知の「アトラクタ」（Probe周辺）に収束させる。これはSamathaの正常な動作だが、Vipassanaが**収束後の状態（S*）のみ**を見ているため、「収束前にどこにいたか」という重要な情報が失われている。

```
[ID入力]   → Vitakka → s0（Probe近く）  → Vicara → S*（Probe近く）
[OOD入力]  → Vitakka → s0（Probe遠い）  → Vicara → S*（Probe近く）← 収束により情報損失
```

## Proposed Solution: Grounding Metrics

Vipassanaに「Grounding Metrics」を追加し、収束前の状態（s0またはz）からの情報を活用する。

### Feature A: Semantic Drift
1. **s0_min_dist**: s0からProbeへの最小距離（収束前のOOD度）
2. **drift_magnitude**: `||S* - s0||`（収束による移動量）

```python
# 提案する実装
def _compute_grounding_features(
    self, s_star: torch.Tensor, s0: torch.Tensor, probes: torch.Tensor
) -> Tuple[torch.Tensor, torch.Tensor]:
    # s0からProbeへの距離（OOD度）
    s0_dists = compute_pairwise_distances(s0, probes)
    s0_min_dist, _ = torch.min(s0_dists, dim=1, keepdim=True)
    
    # 収束による移動量
    drift_magnitude = torch.norm(s_star - s0, dim=1, keepdim=True)
    
    return s0_min_dist, drift_magnitude
```

### Feature B: Reconstruction Reality Check（将来拡張）
Adapter出力zをReconstructionし、入力との乖離を測定。

## Changes Required

1. **Vipassana forward signature変更**: s0を引数に追加
2. **VipassanaEngine変更**: SamathaからVipassanaへs0を渡す
3. **Trust Head入力拡張**: 4 → 6 features
   - 既存: velocity, avg_energy, s_star_min_dist, entropy
   - 追加: s0_min_dist, drift_magnitude
4. **SantanaLog拡張**: s0をメタデータとして保存

## Expected Behavior

```
[ID入力]:  s0_min_dist小, drift小 → Trust Score高
[OOD入力]: s0_min_dist大, drift大 → Trust Score低
```

## Related Files

- `satipatthana/components/vipassana/standard.py`
- `satipatthana/core/engines.py`
- `satipatthana/core/system.py`

## Labels

- enhancement
- vipassana
- ood-detection

**Comment by tamatebox** [comment]
## ✅ Implementation Complete

Grounding Metrics と追加のOOD検出機能を実装しました。

### 実装内容

#### 1. Grounding Metrics (Feature A) - Completed
- **s0_min_dist**: s0からProbeへの最小距離（収束前のOOD度）
- **drift_magnitude**: `||S* - s0||`（収束による移動量）

これにより、Vicaraの収束で失われる「収束前にどこにいたか」の情報を保持。

#### 2. Reconstruction Error (Feature B) - Completed
- **recon_error**: 入力の再構成誤差（高い = OOD/ハルシネーション）
- `samatha_recon_head` を使用して自動計算

#### 3. Void Path Training - Completed
OOD検出能力を学習させるため、Stage 2に **Void Path** を追加：

```python
from satipatthana.data import FilteredNoiseVoid

# 学習データから離れたOODサンプルを生成
void_dataset = FilteredNoiseVoid(
    reference_data=train_data,
    shape=(input_dim,),
    length=1000,
    min_distance=0.3,  # 学習データから0.3以上離れた点のみ
)

trainer = SatipatthanaTrainer(
    ...,
    void_dataset=void_dataset,  # Stage 2で使用
)
```

### Trust Head Features (6 → 7)

| Feature | Source | Purpose |
|---------|--------|---------|
| velocity | Trajectory | 収束速度 |
| avg_energy | Trajectory | 軌跡の滑らかさ |
| min_dist | S* | Probeへの距離（既存） |
| entropy | S* | Probe分布のエントロピー |
| **s0_min_dist** | s0 | **収束前のOOD度** ✨ |
| **drift_magnitude** | s0, S* | **収束による移動量** ✨ |
| **recon_error** | Recon | **再構成誤差** ✨ |

### Stage 2 Training: 5-way Split

```
Clean Path     → Target: 1.0 (高信頼)
Augmented Path → Target: 1.0 - severity
Drunk Path     → Target: 0.0 (内部異常)
Mismatch Path  → Target: 0.0 (論理不整合)
Void Path      → Target: 0.0 (OOD入力) ✨ NEW
```

### Files Changed

- `satipatthana/components/vipassana/standard.py` - Grounding Metrics + recon_error
- `satipatthana/core/engines.py` - VipassanaEngine updated
- `satipatthana/core/system.py` - recon_error computation
- `satipatthana/train/trainer.py` - Void Path + VoidDataset support
- `satipatthana/data/void_dataset.py` - NEW: VoidDataset, FilteredNoiseVoid
- `notebooks/02_trust_score_explained.ipynb` - Updated demo

### Tests

All 301 tests passing ✅
